### PR TITLE
Add some simple PE lit testcases

### PR DESCRIPTION
**Pull request by learn-more**
All testcases compiled with tcc/tcc64
the 'empty' PE files do as less as possible (so no imports, no exports, close to no code, just return 1)
the 'simple' PE files call puts, and the dll's have an additional export (and a bunch of internal symbols exported)

### CODE REVIEW

File: tests/PE/sections/empty-pe32-dll.test [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  - Name:            .text
+    Characteristics: [ IMAGE_SCN_CNT_CODE, IMAGE_SCN_MEM_EXECUTE, IMAGE_SCN_MEM_READ ]
+    VirtualAddress:  4096
+    VirtualSize:     24
+    SectionData:     5589E581EC0000000090B801000000E900000000C9C20C00
+symbols:         []
+...
+
+# CHECK: FILE MAP:
+# CHECK: 000-1a0	        416		[PE Headers]
```
Comment (haberman) [inline tests/PE/sections/empty-pe32-dll.test:80]: For ELF we try to attribute section headers to their corresponding sections. For example, the section header for `.text` is attributed to `.text` instead of `[ELF Headers]`.

Would this be possible for PE?

**Review by haberman — COMMENTED** [review]
These are great, thanks!

File: tests/PE/sections/empty-pe32-dll.test (hunk 1 above)
Comment (learn-more) [reply tests/PE/sections/empty-pe32-dll.test:80]: As outlined in #220, it can be split into:
* DOS Header
* DOS Stub
* Rich Header? (not sure if this can be done reliable, otherwise it will be dumped on top of the DOS Stub)
* NT header
* Per-section header (for each .text, .rdata etc section)

Do you just want the Per-section header here, or also the other DOS/NT header split?

Comment (learn-more) [reply tests/PE/sections/empty-pe32-dll.test:80]: and do you want that only for `segments`, or also for `sections` ?

Comment (haberman) [reply tests/PE/sections/empty-pe32-dll.test:80]: I think mainly the per-section headers should be attributed to their respective sections. The other headers sound like they are not specific to any particular section, so `[PE Headers]` sounds like a good fallback for them.

For ELF I've recently changed segments to work the same way: each segment header is attributed to its respective segment.

Comment (learn-more) [reply tests/PE/sections/empty-pe32-dll.test:80]: Option 1:
```
Maps for empty-pe64-exe.test.tmp.obj:

FILE MAP:
000-188         392             [PE Headers]
188-1b0          40             .text Header
1b0-1d8          40             .pdata Header
1d8-200          40             [Unmapped]
200-400         512             .text
400-600         512             .pdata

VM MAP:
0000-0188               392             [PE Headers]
0188-01b0                40             .text Header
01b0-01d8                40             .pdata Header
01d8-1000              3624             [-- Nothing mapped --]
1000-1020                32             .text
1020-2000              4064             [-- Nothing mapped --]
2000-200c                12             .pdata
```

Or option 2:
```
Maps for empty-pe64-exe.test.tmp.obj:

FILE MAP:
000-040          64             DOS Header
040-080          64             DOS Stub
080-188         264             NT Header
188-1b0          40             .text Header
1b0-1d8          40             .pdata Header
1d8-200          40             [Unmapped]
200-400         512             .text
400-600         512             .pdata

VM MAP:
0000-0040                64             DOS Header
0040-0080                64             DOS Stub
0080-0188               264             NT Header
0188-01b0                40             .text Header
01b0-01d8                40             .pdata Header
01d8-1000              3624             [-- Nothing mapped --]
1000-1020                32             .text
1020-2000              4064             [-- Nothing mapped --]
2000-200c                12             .pdata
```

Comment (haberman) [reply tests/PE/sections/empty-pe32-dll.test:80]: Thanks for trying these out. After playing with the ELF support, I've change my mind on this (sorry for the change). I think I want to change ELF to use:

- [ELF Header] (for the fixed-size ELF header at the top of the file)
- [ELF Section Headers] (used for all section headers)
- [ELF Program Headers] (used for all program headers)

For PE I think this would mean:

- [PE Headers] (for DOS and NT headers)
- [PE Section Headers] (for all section headers)
- [PE Image Headers] (for the "optional" header, only present in images)

This would be similar to your "option 1", except collapse `.text Header` and `.pdata Header` into `[PE Section Headers]`.

Rationale: each individual section header is not big enough to be interesting, but seeing them all grouped together under `[PE Section Headers]` can help users know how much file space in total they are spending for having N sections in their file.

File: src/pe.cc [hunk 2]
```diff
@@ -12,56 +12,56 @@
...
+constexpr size_t kNameSize = 8;
+constexpr size_t kSymbol16Size = 18;
+constexpr size_t kSymbol32Size = 20;
+constexpr size_t kSectionSize = 40;
+constexpr size_t kRelocationSize = 10;
+constexpr size_t kBaseRelocationBlockSize = 8;
+constexpr size_t kImportDirectoryTableEntrySize = 20;
+constexpr size_t kResourceDirectoryTableSize = 16;
+constexpr size_t kResourceDirectoryEntriesSize = 8;
+constexpr size_t kResourceDataEntrySize = 16;
```
Comment (learn-more) [inline src/pe.cc:39]: @haberman I hope you don't mind, I changed these as well.

File: src/pe.cc [hunk 3]
```diff
@@ -109,23 +110,37 @@ bool PeFile::Initialize() {
     return false;
   }
 
-  string_view exe_header =
-      GetRegion(dos_header_.AddressOfNewExeHeader, sizeof(pe_header));
-  memcpy(&pe_header_, exe_header.data(), exe_header.size());
+  PE_TYPE Magic;
+  auto pe_end =
+      CheckedAdd(dos_header_.AddressOfNewExeHeader, sizeof(pe_header_));
+  if (CheckedAdd(pe_end, sizeof(Magic)) > data_.size()) {
```
Comment (learn-more) [inline src/pe.cc:116]: @haberman I changed this code, because I also wanted to check the magic field of the optional header following the PE header.
If I understood the problem from fuzzing correctly this should still work fine?

Would it be possible for you to add a yaml file for the 'bad' PE file that the fuzzer produced?

File: src/pe.cc [hunk 4]
```diff
@@ -109,23 +110,37 @@ bool PeFile::Initialize() {
...
-  // TODO(mj): Parse PE header further to determine this
-  is_64bit_ = false;
+  memcpy(&Magic, data_.data() + pe_end, sizeof(Magic));
+
+  if (Magic != PE_TYPE::PE32 && Magic != PE_TYPE::PE32_PLUS) {
+    // Unknown PE magic
+    return false;
+  }
+
+  is_64bit_ = Magic == PE_TYPE::PE32_PLUS;
```
Comment (learn-more) [inline src/pe.cc:137]: Some structs are different on 64 bit (none of those are used yet, but I wanted to add this check in advance)

**Review by learn-more — COMMENTED** [review]
Some notes for your consideration.

File: src/pe.cc (hunk 2 above)
Comment (haberman) [reply src/pe.cc:39]: Yep np!

File: src/pe.cc (hunk 3 above)
Comment (haberman) [reply src/pe.cc:116]: Every time we fix a fuzzing bug, the payload that crashed the program is added to the [fuzz corpus](https://github.com/google/bloaty/tree/master/tests/testdata/fuzz_corpus). The [`fuzz_test` test](https://github.com/google/bloaty/blob/master/CMakeLists.txt#L382-L384) will always test against every payload in the corpus.

If the code is still passing `fuzz_test`, we should be good!

We don't specify these payloads with yaml, because often they aren't valid input files at all, so `yaml2obj` wouldn't be capable of producing them.

File: tests/PE/sections/empty-pe32-dll.test [hunk 5]
```diff
@@ -0,0 +1,90 @@
+# RUN: %yaml2obj %s -o %t.obj
+# RUN: %bloaty --raw-map %t.obj | %FileCheck %s --dump-input fail
+
+--- !COFF
+OptionalHeader:
```
Comment (haberman) [inline tests/PE/sections/empty-pe32-dll.test:5]: How is this optional header reflected in the map?

I was hoping to report these as `[PE Image Headers]` or similar.

File: src/pe.cc (hunk 3 above)
Comment (learn-more) [reply src/pe.cc:116]: Great, thanks.

File: tests/PE/sections/empty-pe32-dll.test (hunk 5 above)
Comment (learn-more) [reply tests/PE/sections/empty-pe32-dll.test:5]: It is included in the `[PE Headers]` entry.
For a PE file there are always:
- DOS Header
- Optional DOS stub (between DOS and NT header)
- NT Headers (See IMAGE_OPTIONAL_HEADER32 and IMAGE_OPTIONAL_HEADER64 in winnt.h):
  - Signature
  - FileHeader
  - OptionalHeader (contradicting it's name, it is not optional)

Comment (learn-more) [reply tests/PE/sections/empty-pe32-dll.test:5]: @haberman is that good enough, or do you want me to rename `[PE Headers]` to `[PE Image Headers]`?

Comment (haberman) [reply tests/PE/sections/empty-pe32-dll.test:5]: > OptionalHeader (contradicting it's name, it is not optional)

I don't think that's entirely true: I believe it's required on DLLs and executables, but not on object files. It's the equivalent of what ELF calls "program headers."

In my comment above I was thinking of splitting the optional headers out into something like `[Image Headers]`. But since PE is having a lot more header types overall, it might not be worth splitting this out. You've already split out the section headers, which is the most important part.

So I think this is great as is.
